### ISSUE TITLE
Enabling more compiler warnings by default

### ISSUE THREAD

**Issue by springmeyer** [Project-OSRM/osrm-backend#3164]
We can help prevent common bugs by enabling more compiler warnings and adding `-Werror` to treat warnings as errors (at least on CI systems with consistent compilers).

I'd like to start by proposing adding `-Wsign-compare` and `-Wconversion` and fixing all the warnings they trigger as a first step.

/cc @chaupow @MoKob

**Comment by springmeyer** [comment]
Here is the current output of warnings with `-Wsign-compare -Wconversion` when building with clang++-3.8: https://gist.github.com/springmeyer/1dd8dfa83a2dfce7d2e159502373db5f

**Comment by springmeyer** [comment]
Many warnings for `-Wsign-compare -Wconversion` will come from third party headers (like boost). We want to ensure osrm-backend builds without warnings and suppress those from third party code. To do this I propose:

```c++

#include <osrm/warning_supress_start.hpp>
#include <third party include>
#include <osrm/warning_supress_end.hpp>
```

Where `osrm/warning_supress_start.hpp` looks like:

```c++
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunknown-pragmas" // clang+gcc
#pragma GCC diagnostic ignored "-Wpragmas" // gcc
#pragma GCC diagnostic ignored "-W#pragma-messages"
#pragma GCC diagnostic ignored "-Wsign-compare"
#pragma GCC diagnostic ignored "-Wconversion"
```

And `osrm/warning_supress_end.hpp` looks like:

```c++
#pragma GCC diagnostic pop
```

As we add more default warnings, we can then ensure they are suppressed by adding them to one place: `osrm/warning_supress_start.hpp`

**Comment by chaupow** [comment]
I would also suggest adding `-Wshadow` as we have lots of `typedef` in the code that are potentially shadowing each other

**Comment by springmeyer** [comment]
> Note: the alternative method of suppressing warnings from third party or system headers is to ensure that the include paths for them are referenced not with -I/path/to but with -isystem /path/to (https://gcc.gnu.org/onlinedocs/cpp/System-Headers.html). 

@chaupow and I took a look at the cmake build last week with an eye towards this issue. We found that all paths to third party dependencies are added with `-isystem`. So, this is actually working great for compilers that support `-isystem` (unix only) and there should be no need to use the `#include <osrm/warning_suppress_start.hpp>` method I propose above.

Also, @chaupow has kicked off a first effort at adding a new warning, and fixing the new warnings those arose from it at https://github.com/Project-OSRM/osrm-backend/pull/3209. I propose we close this issue, continue working on https://github.com/Project-OSRM/osrm-backend/pull/3209, and then that is finished we can pick new warnings and create new PRs to work on this.
